modules/sfp_massdns.py

```diff
@@ -0,0 +1,406 @@
+        finally:
+            # delete temporary files
+            Path(resolversFile.name).unlink()
+            Path(subdomainsFile.name).unlink()
+
+        return self.validateHosts(records)
+
+    def sendEvent(self, source, host, method=None):
+        if method is None:
+            method = ""
+        host = host.lower()
+        # skip if we've already sent this event
+        eventDataHash = self.sf.hashstring(host)
+        if eventDataHash in self.state["sent_events"]:
+            self.sf.debug("Skipping already-sent event")
+            return
+        elif eventDataHash in self.state["handled_events"]:
+            self.sf.debug("Not sending already-handled event")
+            return
+        self.state["sent_events"].append(eventDataHash)
+        self.sf.info(f"Found subdomain via {method}: {host}")
+        # Report the host
+        e = SpiderFootEvent("INTERNET_NAME", host, self.__name__, source)
+        self.notifyListeners(e)
+
+    def verifyNameservers(self, nameservers, timeout=2):
+        """Check each resolver to make sure it can actually resolve DNS names
+
+        Args:
+            nameservers (list): nameservers to verify
+            timeout (int): timeout for dns query
+
+        Returns:
+            boolean: whether any of the nameservers are valid
+        """
+        validResolvers = []
+        with self.threadPool(threads=100, name='sfp_massdns_verify_nameservers') as pool:
+            for resolver, error in pool.map(nameservers, self.verifyNameserver):
+                if not error:
+                    validResolvers.append(resolver)
+                else:
+                    self.sf.debug(str(error))
+        return validResolvers
+
+    def verifyNameserver(self, nameserver, timeout=2):
+        """Validate a nameserver by making a sample query and a garbage query
+
+        Args:
+            nameserver (str): nameserver to verify
+            timeout (int): timeout for dns query
+
+        Returns:
+            boolean: whether the nameserver is valid
+        """
+        error = None
+
+        resolver = dns.resolver.Resolver()
+        resolver.timeout = timeout
+        resolver.lifetime = timeout
+        resolver.nameservers = [nameserver]
+
+        # first, make sure it can resolve google.com
+        try:
+            resolver.query("www.google.com", "A")
+        except Exception:
+            error = f"Nameserver {nameserver} failed to resolve basic query within {timeout} seconds."
+
+        # then, make sure it isn't feeding us garbage data
+        randpool = "bcdfghjklmnpqrstvwxyz3456789"
+        randhost = "".join([random.SystemRandom().choice(randpool) for x in range(10)]) + ".google.com"
+        try:
+            results = list(resolver.query(randhost, "A"))
+            if results:
+                error = f"Nameserver {nameserver} returned garbage data."
+        except Exception:
+            # Garbage query to nameserver failed successfully ;)
+            pass
```

Comment:
Outside the scope of this PR, but we should probably augment `checkDnsWildcard()` to also take a `resolver` argument rather than duplicate code here.

https://github.com/smicallef/spiderfoot/blob/0a505a836ba7cd6df6a8e3b60605dcbfc757fd41/sflib.py#L2500-L2519